In [1]:
import pandas as pd
Classification=pd.read_csv(r'C:\Users\sandy\Desktop\Project_realected_practice\Online_shoppers_purchasing\online+shoppers+purchasing+intention+dataset\online_shoppers_intention.csv')
Classification.head()

Administrative  Administrative_Duration  Informational  \
0               0                      0.0              0   
1               0                      0.0              0   
2               0                      0.0              0   
3               0                      0.0              0   
4               0                      0.0              0   

   Informational_Duration  ProductRelated  ProductRelated_Duration  \
0                     0.0               1                 0.000000   
1                     0.0               2                64.000000   
2                     0.0               1                 0.000000   
3                     0.0               2                 2.666667   
4                     0.0              10               627.500000   

   BounceRates  ExitRates  PageValues  SpecialDay Month  OperatingSystems  \
0         0.20       0.20         0.0         0.0   Feb                 1   
1         0.00       0.10         0.0         0.0   Feb                 2   
2         0.20       0.20         0.0         0.0   Feb                 4   
3         0.05       0.14         0.0         0.0   Feb                 3   
4         0.02       0.05         0.0         0.0   Feb                 3   

   Browser  Region  TrafficType        VisitorType  Weekend  Revenue  
0        1       1            1  Returning_Visitor    False    False  
1        2       1            2  Returning_Visitor    False    False  
2        1       9            3  Returning_Visitor    False    False  
3        2       2            4  Returning_Visitor    False    False  
4        3       1            4  Returning_Visitor     True    False

In [2]:
Classification.shape

(12330, 18)

In [3]:
Classification['VisitorType'].unique()

array(['Returning_Visitor', 'New_Visitor', 'Other'], dtype=object)

In [4]:
Classification['Month'].unique()

array(['Feb', 'Mar', 'May', 'Oct', 'June', 'Jul', 'Aug', 'Nov', 'Sep',
       'Dec'], dtype=object)

In [5]:
Classification['TrafficType'].unique()

array([ 1,  2,  3,  4,  5,  6,  7,  8,  9, 10, 11, 12, 13, 14, 15, 18, 19,
       16, 17, 20])

In [6]:
Classification['Revenue'].unique()

array([False,  True])

In [7]:
Classification['Weekend'].unique()

array([False,  True])

There are 4 column in categorical values need to be do a encoding to change it into a numerical 

In [8]:
# Import label encoder 
from sklearn import preprocessing 
a=['VisitorType','Revenue','Weekend']
label_encoder = preprocessing.LabelEncoder() 
for i in a:
    Classification[i]= label_encoder.fit_transform(Classification[i]) 
    print(Classification[i].unique())

[2 0 1]
[0 1]
[0 1]


so In visitor type 'Returning_Visitor'--> 2, 'New_Visitor'--> 0, 'Other'---> 1

Similarly in Revenue False --> 0 True ---> 1

In Weekend False --> 0 True ---> 1

Now need to map Month since there the mounth is not arranged regularly I am using the map to have a control of which month have which number.

In [9]:

Classification.Month=Classification.Month.map({'Feb':2, 'Mar':3, 'May':5, 'Oct':10, 'June':6, 'Jul':7, 'Aug':8, 'Nov':9, 'Sep':11,'Dec':12})
Classification.head()


Administrative  Administrative_Duration  Informational  \
0               0                      0.0              0   
1               0                      0.0              0   
2               0                      0.0              0   
3               0                      0.0              0   
4               0                      0.0              0   

   Informational_Duration  ProductRelated  ProductRelated_Duration  \
0                     0.0               1                 0.000000   
1                     0.0               2                64.000000   
2                     0.0               1                 0.000000   
3                     0.0               2                 2.666667   
4                     0.0              10               627.500000   

   BounceRates  ExitRates  PageValues  SpecialDay  Month  OperatingSystems  \
0         0.20       0.20         0.0         0.0      2                 1   
1         0.00       0.10         0.0         0.0      2                 2   
2         0.20       0.20         0.0         0.0      2                 4   
3         0.05       0.14         0.0         0.0      2                 3   
4         0.02       0.05         0.0         0.0      2                 3   

   Browser  Region  TrafficType  VisitorType  Weekend  Revenue  
0        1       1            1            2        0        0  
1        2       1            2            2        0        0  
2        1       9            3            2        0        0  
3        2       2            4            2        0        0  
4        3       1            4            2        1        0

In [10]:
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler, LabelEncoder
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import accuracy_score

Seperating the Features and Tragets.

In [11]:
Feature_1=Classification.drop('VisitorType',axis=1)
Feature_1.head()

Administrative  Administrative_Duration  Informational  \
0               0                      0.0              0   
1               0                      0.0              0   
2               0                      0.0              0   
3               0                      0.0              0   
4               0                      0.0              0   

   Informational_Duration  ProductRelated  ProductRelated_Duration  \
0                     0.0               1                 0.000000   
1                     0.0               2                64.000000   
2                     0.0               1                 0.000000   
3                     0.0               2                 2.666667   
4                     0.0              10               627.500000   

   BounceRates  ExitRates  PageValues  SpecialDay  Month  OperatingSystems  \
0         0.20       0.20         0.0         0.0      2                 1   
1         0.00       0.10         0.0         0.0      2                 2   
2         0.20       0.20         0.0         0.0      2                 4   
3         0.05       0.14         0.0         0.0      2                 3   
4         0.02       0.05         0.0         0.0      2                 3   

   Browser  Region  TrafficType  Weekend  Revenue  
0        1       1            1        0        0  
1        2       1            2        0        0  
2        1       9            3        0        0  
3        2       2            4        0        0  
4        3       1            4        1        0

In [12]:
Target=Classification['VisitorType']
Target.head()

0    2
1    2
2    2
3    2
4    2
Name: VisitorType, dtype: int64

Scaling the Features since it is a classification kind of problem no need for scale the target

Note: Standradization should be done after splitted the data in here I have mistakenly standadized before 

In [13]:
from sklearn.preprocessing import StandardScaler
# Standardize the features
scaler = StandardScaler()
Feature = scaler.fit_transform(Feature_1)
Feature

array([[-0.69699296, -0.4571914 , -0.3964779 , ..., -0.76262903,
        -0.55055169, -0.42787179],
       [-0.69699296, -0.4571914 , -0.3964779 , ..., -0.51418219,
        -0.55055169, -0.42787179],
       [-0.69699296, -0.4571914 , -0.3964779 , ..., -0.26573535,
        -0.55055169, -0.42787179],
       ...,
       [-0.69699296, -0.4571914 , -0.3964779 , ...,  2.21873304,
         1.81635987, -0.42787179],
       [ 0.50722805, -0.03291592, -0.3964779 , ...,  1.72183936,
        -0.55055169, -0.42787179],
       [-0.69699296, -0.4571914 , -0.3964779 , ..., -0.51418219,
         1.81635987, -0.42787179]])

Spliting the Data for training and testing.

In [14]:
X_train, X_test, y_train, y_test = train_test_split(Feature, Target, test_size=0.3, random_state=10,stratify=Target)

In [15]:
y_test.value_counts()

VisitorType
2    3165
0     508
1      26
Name: count, dtype: int64

In [16]:
y_train.value_counts()

VisitorType
2    7386
0    1186
1      59
Name: count, dtype: int64

Now training the LogisticRegression model 

In [17]:

Logi_model = LogisticRegression(max_iter=10000)
Logi_model.fit(X_train, y_train)

# Predict and evaluate
Logi_model_pred = Logi_model.predict(X_test)
accuracy = accuracy_score(y_test, Logi_model_pred)
print(f"Accuracy: {accuracy}")

# Print the classification report
from sklearn.metrics import classification_report
print(classification_report(y_test, Logi_model_pred ))
crosstab_result = pd.crosstab(y_test,Logi_model_pred )

print(crosstab_result)

Accuracy: 0.858610435252771
              precision    recall  f1-score   support

           0       0.49      0.09      0.16       508
           1       0.79      0.58      0.67        26
           2       0.87      0.98      0.92      3165

    accuracy                           0.86      3699
   macro avg       0.72      0.55      0.58      3699
weighted avg       0.82      0.86      0.82      3699

col_0         0   1     2
VisitorType              
0            47   0   461
1             1  15    10
2            47   4  3114


If we want to increase the model acurracy means we want to drop the 0 - class since it have a less f1-score but I will leave it as it is and I will move to the next algorithm to check which algorithm gives a best accuracy

KNN- K-Neareast Neighbour

In [18]:
from sklearn.metrics import accuracy_score
from sklearn.metrics import classification_report
from sklearn.neighbors import KNeighborsClassifier
for k in range(1,10):
    KNN_model = KNeighborsClassifier(n_neighbors = k).fit(X_train,y_train)
    Knn_pred = KNN_model.predict(X_test)
    final=accuracy_score(y_test, Knn_pred)
    print(f'Accuracy score of k-value {k} is {final}')
    final1=classification_report(y_test, Knn_pred)
    print(f'Accuracy score of k-value {k} is {final1}')

Accuracy score of k-value 1 is 0.7961611246282779
Accuracy score of k-value 1 is               precision    recall  f1-score   support

           0       0.28      0.30      0.29       508
           1       0.60      0.58      0.59        26
           2       0.88      0.88      0.88      3165

    accuracy                           0.80      3699
   macro avg       0.59      0.58      0.59      3699
weighted avg       0.80      0.80      0.80      3699

Accuracy score of k-value 2 is 0.741551770748851
Accuracy score of k-value 2 is               precision    recall  f1-score   support

           0       0.26      0.47      0.34       508
           1       0.47      0.62      0.53        26
           2       0.90      0.79      0.84      3165

    accuracy                           0.74      3699
   macro avg       0.55      0.62      0.57      3699
weighted avg       0.81      0.74      0.77      3699

Accuracy score of k-value 3 is 0.8302243849689105
Accuracy score of k-value 3

In this the best accuracy score is 84.6 which is achieved in k value k=7 when comparing with 7 nearseast neighbour it performs well

Now lets go with Descion tree and random forest and check its accuracy score 

In [19]:
from sklearn import tree
clf = tree.DecisionTreeClassifier(random_state=0)
clf.fit(X_train,y_train)
y_test_pred_DT = clf.predict(X_test)
final=accuracy_score(y_test, y_test_pred_DT)
print(f'Accuracy score is {final}')
final1=classification_report(y_test, y_test_pred_DT)
print(f'Accuracy score is {final1}')

Accuracy score is 0.8318464449851312
Accuracy score is               precision    recall  f1-score   support

           0       0.40      0.39      0.40       508
           1       0.60      0.58      0.59        26
           2       0.90      0.90      0.90      3165

    accuracy                           0.83      3699
   macro avg       0.63      0.62      0.63      3699
weighted avg       0.83      0.83      0.83      3699



We can optimize this as well but let see how random forest and XGBoost works

In [20]:
# import Random Forest classifier
from sklearn.ensemble import RandomForestClassifier
from sklearn.metrics import accuracy_score
from sklearn.metrics import classification_report

Rfc = RandomForestClassifier(n_estimators=100,random_state=0)
Rfc.fit(X_train, y_train)
Rfc_ypred= Rfc.predict(X_test)

print('Model accuracy score with 100 decision-trees : {0:0.4f}'. format(accuracy_score(y_test,Rfc_ypred)))
print(classification_report(y_test,Rfc_ypred))


Model accuracy score with 100 decision-trees : 0.8746
              precision    recall  f1-score   support

           0       0.59      0.36      0.45       508
           1       0.88      0.58      0.70        26
           2       0.90      0.96      0.93      3165

    accuracy                           0.87      3699
   macro avg       0.79      0.63      0.69      3699
weighted avg       0.86      0.87      0.86      3699



In [21]:
from sklearn.ensemble import RandomForestClassifier
from sklearn.model_selection import GridSearchCV
from sklearn.metrics import accuracy_score, classification_report

# Define the Random Forest classifier
rfc = RandomForestClassifier(random_state=0)

# Define the hyperparameters and their corresponding values to search over
param_grid = {
    'n_estimators': [50, 100, 200],         # Number of trees in the forest
    'max_features': ['auto', 'sqrt', 'log2'], # Number of features to consider at every split
    'max_depth': [None, 10, 20, 30],        # Maximum depth of the tree
    'min_samples_split': [2, 5, 10],        # Minimum number of samples required to split a node
    'min_samples_leaf': [1, 2, 4],          # Minimum number of samples required at each leaf node
    'bootstrap': [True, False]              # Whether bootstrap samples are used when building trees
}

# Initialize GridSearchCV with the Random Forest classifier and the hyperparameters
grid_search = GridSearchCV(estimator=rfc, 
                           param_grid=param_grid, 
                           cv=5, 
                           n_jobs=-1, 
                           verbose=2, 
                           scoring='accuracy')

# Fit GridSearchCV on the training data
grid_search.fit(X_train, y_train)

# Get the best hyperparameters from Grid Search
best_params = grid_search.best_params_
print(f"Best hyperparameters: {best_params}")

# Use the best estimator to make predictions
best_rfc = grid_search.best_estimator_
y_pred = best_rfc.predict(X_test)

# Evaluate the model
print('Model accuracy score with best parameters: {0:0.4f}'.format(accuracy_score(y_test, y_pred)))
print(classification_report(y_test, y_pred))


Fitting 5 folds for each of 648 candidates, totalling 3240 fits


c:\Users\sandy\AppData\Local\Programs\Python\Python312\Lib\site-packages\sklearn\model_selection\_validation.py:540: FitFailedWarning: 
1080 fits failed out of a total of 3240.
The score on these train-test partitions for these parameters will be set to nan.
If these failures are not expected, you can try to debug them by setting error_score='raise'.

Below are more details about the failures:
--------------------------------------------------------------------------------
407 fits failed with the following error:
Traceback (most recent call last):
  File "c:\Users\sandy\AppData\Local\Programs\Python\Python312\Lib\site-packages\sklearn\model_selection\_validation.py", line 888, in _fit_and_score
    estimator.fit(X_train, y_train, **fit_params)
  File "c:\Users\sandy\AppData\Local\Programs\Python\Python312\Lib\site-packages\sklearn\base.py", line 1466, in wrapper
    estimator._validate_params()
  File "c:\Users\sandy\AppData\Local\Programs\Python\Python312\Lib\site-packages\sklearn\ba

Best hyperparameters: {'bootstrap': False, 'max_depth': 10, 'max_features': 'sqrt', 'min_samples_leaf': 1, 'min_samples_split': 2, 'n_estimators': 200}
Model accuracy score with best parameters: 0.8732
              precision    recall  f1-score   support

           0       0.59      0.31      0.40       508
           1       0.94      0.58      0.71        26
           2       0.89      0.97      0.93      3165

    accuracy                           0.87      3699
   macro avg       0.81      0.62      0.68      3699
weighted avg       0.85      0.87      0.86      3699



In [24]:
y_pred = best_rfc.predict(X_test)
print('Model accuracy score with best parameters: {0:0.4f}'.format(accuracy_score(y_test, y_pred)))
print(classification_report(y_test, y_pred))

Model accuracy score with best parameters: 0.8732
              precision    recall  f1-score   support

           0       0.59      0.31      0.40       508
           1       0.94      0.58      0.71        26
           2       0.89      0.97      0.93      3165

    accuracy                           0.87      3699
   macro avg       0.81      0.62      0.68      3699
weighted avg       0.85      0.87      0.86      3699



In [21]:
!pip install xgboost

^C



[notice] A new release of pip is available: 24.0 -> 24.2
[notice] To update, run: python.exe -m pip install --upgrade pip


Now lets Try with the XGBoost classifier algorithm

In [ ]:
# Import XGBoost classifier
from xgboost import XGBClassifier
from sklearn.metrics import accuracy_score, classification_report

# Initialize the XGBoost model
xgb_model = XGBClassifier(n_estimators=100, random_state=0, use_label_encoder=False, eval_metric='mlogloss')

# Fit the model to the training data
xgb_model.fit(X_train, y_train)

# Predict on the test data
xgb_ypred = xgb_model.predict(X_test)

# Evaluate the model
print('Model accuracy score with 100 estimators: {0:0.4f}'.format(accuracy_score(y_test, xgb_ypred)))
print(classification_report(y_test, xgb_ypred))


c:\Users\sandy\AppData\Local\Programs\Python\Python312\Lib\site-packages\xgboost\core.py:158: UserWarning: [16:11:31] WARNING: C:\buildkite-agent\builds\buildkite-windows-cpu-autoscaling-group-i-0015a694724fa8361-1\xgboost\xgboost-ci-windows\src\learner.cc:740: 
Parameters: { "use_label_encoder" } are not used.

  warnings.warn(smsg, UserWarning)


Model accuracy score with 100 estimators: 0.8751
              precision    recall  f1-score   support

           0       0.58      0.43      0.49       508
           1       0.83      0.58      0.68        26
           2       0.91      0.95      0.93      3165

    accuracy                           0.88      3699
   macro avg       0.77      0.65      0.70      3699
weighted avg       0.86      0.88      0.87      3699



I have tried most of the classifier algorithm the maximum achievable accuracy is 87.5% in a XGBoost so I will keep it as it is and do a  optimization in decision and RandomForest by parameter tuning - Grid search CV let's try at what extent I can increase it accuracy

First I will try with decision tree

In [ ]:
from sklearn.model_selection import RandomizedSearchCV, GridSearchCV
params = {'max_depth': [i for i in range(1,13)],
        'min_samples_split': [i for i in range(1,6)],
        'min_samples_leaf': [i for i in range(1,6)]}


clf = tree.DecisionTreeClassifier()
gcv = GridSearchCV(estimator=clf,param_grid=params)
gcv.fit(X_train,y_train)

Optimized_DT = gcv.best_estimator_
print(gcv.best_estimator_)
Optimized_DT.fit(X_train,y_train)
y_train_pred_Optim_DT = Optimized_DT.predict(X_train)
y_test_pred_Optim_DT = Optimized_DT.predict(X_test)


print(f'Train score {accuracy_score(y_train_pred_Optim_DT,y_train)}')
print(f'Test score {accuracy_score(y_test_pred_Optim_DT,y_test)}')

DecisionTreeClassifier(max_depth=8, min_samples_leaf=5, min_samples_split=4)
Train score 0.8868033831537481
Test score 0.851581508515815


c:\Users\sandy\AppData\Local\Programs\Python\Python312\Lib\site-packages\sklearn\model_selection\_validation.py:540: FitFailedWarning: 
300 fits failed out of a total of 1500.
The score on these train-test partitions for these parameters will be set to nan.
If these failures are not expected, you can try to debug them by setting error_score='raise'.

Below are more details about the failures:
--------------------------------------------------------------------------------
300 fits failed with the following error:
Traceback (most recent call last):
  File "c:\Users\sandy\AppData\Local\Programs\Python\Python312\Lib\site-packages\sklearn\model_selection\_validation.py", line 888, in _fit_and_score
    estimator.fit(X_train, y_train, **fit_params)
  File "c:\Users\sandy\AppData\Local\Programs\Python\Python312\Lib\site-packages\sklearn\base.py", line 1466, in wrapper
    estimator._validate_params()
  File "c:\Users\sandy\AppData\Local\Programs\Python\Python312\Lib\site-packages\sklearn\bas

Conclusion is the XGBOOST model performs well in this its accuracy score is 87.5% so for further deployment I use this model and I will preserve this model for the future use

In [ ]:
import pickle


# Assuming your model is stored in a variable called 'model'
with open('Shoppers_Class.pkl', 'wb') as file:
   pickle.dump(xgb_model, file)


with open('Shoppers_Class.pkl', 'rb') as file:
   pickle.load(file)


Now using the same dataset for clustering

In [26]:
import numpy as np
import pandas as pd

from sklearn.cluster import KMeans


import warnings
warnings.filterwarnings('ignore')


Clustering using Kmeans algorithm

In [27]:

algorithm = (KMeans(n_clusters = 3 ,init='k-means++', n_init = 10 ,max_iter=300,
                       tol=0.0001,  random_state= 111  , algorithm='elkan') )
algorithm.fit(Feature)

y_kmeans = algorithm.fit_predict(Feature)
Feature_1['cluster'] = pd.DataFrame(y_kmeans)
Feature_1.head()

Administrative  Administrative_Duration  Informational  \
0               0                      0.0              0   
1               0                      0.0              0   
2               0                      0.0              0   
3               0                      0.0              0   
4               0                      0.0              0   

   Informational_Duration  ProductRelated  ProductRelated_Duration  \
0                     0.0               1                 0.000000   
1                     0.0               2                64.000000   
2                     0.0               1                 0.000000   
3                     0.0               2                 2.666667   
4                     0.0              10               627.500000   

   BounceRates  ExitRates  PageValues  SpecialDay  Month  OperatingSystems  \
0         0.20       0.20         0.0         0.0      2                 1   
1         0.00       0.10         0.0         0.0      2                 2   
2         0.20       0.20         0.0         0.0      2                 4   
3         0.05       0.14         0.0         0.0      2                 3   
4         0.02       0.05         0.0         0.0      2                 3   

   Browser  Region  TrafficType  Weekend  Revenue  cluster  
0        1       1            1        0        0        2  
1        2       1            2        0        0        1  
2        1       9            3        0        0        2  
3        2       2            4        0        0        2  
4        3       1            4        1        0        1

In [25]:
Feature_1['cluster'].value_counts()

cluster
1    9608
0    1671
2    1051
Name: count, dtype: int64

In [ ]:
import pickle


# Assuming your model is stored in a variable called 'model'
with open('Shoppers_Cluster.pkl', 'wb') as file:
   pickle.dump(algorithm, file)


with open('Shoppers_Cluster.pkl', 'rb') as file:
   pickle.load(file)
